core/trino-spi/src/main/java/io/trino/spi/exchange/ExchangeSink.java

```diff
@@ -46,11 +46,15 @@
 
     /**
      * Notifies the exchange sink that no more data will be appended
+     *
+     * @return future that will be resolved when the finish operation either succeeds or fails
      */
-    void finish();
+    CompletableFuture<?> finish();
 
     /**
      * Notifies the exchange that the write operation has been aborted
+     *
+     * @return future that will be resolved when the abort operation either succeeds or fails
```

Comment:
What should `abort` do when `finish` is already running?

Reply:
> What should abort do when finish is already running?

I think it can be implementation specific. The implementation may decide to keep the `finish` running, or may decided to cancel `finish` and abort. It doesn't really make a difference from the engine perspective.

> What should abort do when another abort is running?

Same here. It is implementation specific. As long as the sink is properly invalidated the engine doesn't really care what happens underneath as the task is already aborted / failed anyway. Regardless I guess it is better to make the `abort` method idempotent. I will change it to first transition the buffer to the aborted state and then call the `ExchangeSink#abort`